server/src/main/java/io/crate/planner/optimizer/joinorder/JoinGraph.java

```diff
@@ -0,0 +1,271 @@
+ * | t1 |--t1.x = t3.z--| t3 |--t3.z = t2.y--| t2 |
+ * +----+               +----+               +----+
+ *</pre>
+ *
+ * <p>
+ * Edges are created and indexed for each equi-join condition
+ * from both directions so a.x = b.y becomes:
+ * </p>
+ * <pre>
+ * a -> Edge[b, a.x, b.y]
+ * b -> Edge[a, a.x, b.y]
+ * </pre>
+ */
+public record JoinGraph(
+    LogicalPlan root,
+    List<LogicalPlan> nodes,
+    Map<LogicalPlan, Set<Edge>> edges,
+    List<Symbol> filters,
+    boolean hasCrossJoin) {
+
+    public record Edge(LogicalPlan to, Symbol left, Symbol right) {}
+
+    JoinGraph joinWith(LogicalPlan root, JoinGraph other) {
+        for (var node : other.nodes) {
+            assert !edges.containsKey(node) : "LogicalPlan" + node + " can't be in both graphs";
+        }
+
+        var newNodes = Lists2.concat(this.nodes, other.nodes);
+        var newEdges = Maps.merge(this.edges, other.edges, Sets::union);
+        var newFilters = Lists2.concat(this.filters, other.filters);
+        var hasCrossJoin = this.hasCrossJoin || other.hasCrossJoin();
+
+        return new JoinGraph(
+            root,
+            newNodes,
+            newEdges,
+            newFilters,
+            hasCrossJoin
+        );
+    }
+
+    JoinGraph withEdges(Map<LogicalPlan, Set<Edge>> edges) {
+        var newEdges = Maps.merge(this.edges, edges, Sets::union);
+        return new JoinGraph(this.root, this.nodes, newEdges, this.filters, this.hasCrossJoin);
+    }
+
+    JoinGraph withFilters(List<Symbol> filters) {
+        var newFilters = Lists2.concat(this.filters, filters);
+        return new JoinGraph(this.root, this.nodes, edges, newFilters, this.hasCrossJoin);
+    }
+
+    JoinGraph withCrossJoin(boolean hasCrossJoin) {
+        return new JoinGraph(this.root, this.nodes, edges, filters, hasCrossJoin);
+    }
+
+    public int size() {
+        return nodes.size();
+    }
+
+    public Set<Edge> edges(LogicalPlan node) {
+        var result = edges.get(node);
+        if (result == null) {
+            return Set.of();
+        }
+        return result;
+    }
+
+    public static JoinGraph create(LogicalPlan plan, Function<LogicalPlan, LogicalPlan> resolvePlan) {
+        return plan.accept(new GraphBuilder(resolvePlan), new GraphBuilder.Context());
+    }
+
+    private static class GraphBuilder extends LogicalPlanVisitor<GraphBuilder.Context, JoinGraph> {
+
+        private static class Context {
+            final Map<Symbol, LogicalPlan> referencesToRelations = new HashMap<>();
+            final Map<RelationName, FieldResolver> resolver = new HashMap<>();
+        }
+
+        private final Function<LogicalPlan, LogicalPlan> resolvePlan;
+
+        GraphBuilder(Function<LogicalPlan, LogicalPlan> resolvePlan) {
+            this.resolvePlan = resolvePlan;
+        }
+
+        @Override
+        public JoinGraph visitPlan(LogicalPlan logicalPlan, Context context) {
+            if (logicalPlan instanceof ForwardingLogicalPlan f) {
```

Comment:
Is this safe to use `ForwardingLogicalPlan` as marker (for what?) here?

Reply:
I added this to process `RootRelationBoundary` and get the original source, I will to double-check if this can cause any trouble.